File: fathersDay/Dsa/Day 31 (230. Kth Smallest Element in a BST)/Solution.java
import java.util.ArrayList;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) { this.val = val; }
}

class Solution {
    public int KthSmallest(TreeNode root, int k) {
        List<Integer> sortedElements = new ArrayList<>();
        inorder(root, sortedElements);
        if (k > 0 && k <= sortedElements.size()) {
            return sortedElements.get(k - 1);
        }
        return -1; // Or throw an exception
    }

    private void inorder(TreeNode node, List<Integer> list) {
        if (node == null) return;
        inorder(node.left, list);
        list.add(node.val);
        inorder(node.right, list);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(1);
        root.right = new TreeNode(4);
        root.left.right = new TreeNode(2);
        Solution solution = new Solution();
        int k = 1;
        int result = solution.KthSmallest(root, k);
        System.out.println(result); // Output: 1
    }
}

File: fathersDay/Dsa/Day 31 (230. Kth Smallest Element in a BST)/Solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    void inorder(TreeNode* node, std::vector<int>& result) {
        if (!node) return;
        inorder(node->left, result);
        result.push_back(node->val);
        inorder(node->right, result);
    }

    int KthSmallest(TreeNode* root, int k) {
        std::vector<int> sorted_elements;
        inorder(root, sorted_elements);
        if (k > 0 && k <= sorted_elements.size()) {
            return sorted_elements[k - 1];
        }
        return -1; // Or throw an exception
    }
};

int main() {
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(1);
    root->right = new TreeNode(4);
    root->left->right = new TreeNode(2);
    Solution solution;
    int k = 1;
    int result = solution.KthSmallest(root, k);
    std::cout << result << std::endl; // Output: 1
    return 0;
}
